airbyte-integrations/connectors/source-pipedrive/source_pipedrive/streams.py

```diff
@@ -0,0 +1,192 @@
+# SOFTWARE.
+#
+
+from abc import ABC
+from typing import Any, Iterable, Mapping, MutableMapping, Optional
+
+import pendulum
+import requests
+from airbyte_cdk.sources.streams.http import HttpStream
+
+PIPEDRIVE_URL_BASE = "https://api.pipedrive.com/v1/"
+
+
+class PipedriveStream(HttpStream, ABC):
+    url_base = PIPEDRIVE_URL_BASE
+    primary_key = "id"
+    data_field = "data"
+
+    def __init__(self, api_token: str, **kwargs):
+        super().__init__(**kwargs)
+        self._api_token = api_token
+
+    def path(self, **kwargs) -> str:
+        class_name = self.__class__.__name__
+        return f"{class_name[0].lower()}{class_name[1:]}"
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return {"api_token": self._api_token}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        :return an iterable containing each record in the response
+        """
+        records = response.json().get(self.data_field) or []
+        yield from records
+
+
+class PaginatedPipedriveStream(PipedriveStream):
+    limit = 50
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        """
+        :param response: the most recent response from the API
+        :return If there is another page in the result, a mapping (e.g: dict) containing information needed to query
+                the next page in the response.
+                If there are no more pages in the result, return None.
+        """
+        pagination_data = response.json().get("additional_data", {}).get("pagination", {})
+        if pagination_data.get("more_items_in_collection") and pagination_data.get("start") is not None:
+            start = pagination_data.get("start") + self.limit
+            return {"start": start}
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = super().request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
+        next_page_token = next_page_token or {}
+        params.update({"limit": self.limit, **next_page_token})
+        return params
+
+
+class IncrementalPipedriveStream(PaginatedPipedriveStream, ABC):
+    cursor_field = "update_time"
+
+    def __init__(self, replication_start_date: pendulum.datetime, **kwargs):
+        super().__init__(**kwargs)
+        self._replication_start_date = replication_start_date
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
```

Comment:
this just proves the point above, what if we will have nonepaginated and incremental stream?